src/test/mocks/MockGsmFailedSellAssetRemainingGhoBalance.sol

```diff
@@ -0,0 +1,582 @@
+   * @dev Buys an underlying asset with GHO
+   * @param originator The originator of the request
+   * @param minAmount The minimum amount of the underlying asset desired for purchase
+   * @param receiver The recipient address of the underlying asset being purchased
+   * @return The amount of underlying asset bought
+   * @return The amount of GHO sold by the user
+   */
+  function _buyAsset(
+    address originator,
+    uint256 minAmount,
+    address receiver
+  ) internal returns (uint256, uint256) {
+    (
+      uint256 assetAmount,
+      uint256 ghoSold,
+      uint256 grossAmount,
+      uint256 fee
+    ) = _calculateGhoAmountForBuyAsset(minAmount);
+
+    _beforeBuyAsset(originator, assetAmount, receiver);
+
+    require(assetAmount > 0, 'INVALID_AMOUNT');
+    require(_currentExposure >= assetAmount, 'INSUFFICIENT_AVAILABLE_EXOGENOUS_ASSET_LIQUIDITY');
+
+    _currentExposure -= uint128(assetAmount);
+    _accruedFees += fee.toUint128();
+    IGhoToken(GHO_TOKEN).transferFrom(originator, address(this), ghoSold);
+    IGhoToken(GHO_TOKEN).burn(grossAmount);
+    IERC20(UNDERLYING_ASSET).safeTransfer(receiver, assetAmount);
+
+    emit BuyAsset(originator, receiver, assetAmount, ghoSold, fee);
+    return (assetAmount, ghoSold);
+  }
+
+  /**
+   * @dev Hook that is called before `buyAsset`.
+   * @dev This can be used to add custom logic
+   * @param originator Originator of the request
+   * @param amount The amount of the underlying asset desired for purchase
+   * @param receiver Recipient address of the underlying asset being purchased
+   */
+  function _beforeBuyAsset(address originator, uint256 amount, address receiver) internal virtual {}
+
+  /**
+   * @dev Sells an underlying asset for GHO
+   * @param originator The originator of the request
+   * @param maxAmount The maximum amount of the underlying asset desired to sell
+   * @param receiver The recipient address of the GHO being purchased
+   * @return The amount of underlying asset sold
+   * @return The amount of GHO bought by the user
+   */
+  function _sellAsset(
+    address originator,
+    uint256 maxAmount,
+    address receiver
+  ) internal returns (uint256, uint256) {
+    (
+      uint256 assetAmount,
+      uint256 ghoBought,
+      uint256 grossAmount,
+      uint256 fee
+    ) = _calculateGhoAmountForSellAsset(maxAmount);
+
+    _beforeSellAsset(originator, assetAmount, receiver);
+
+    require(assetAmount > 0, 'INVALID_AMOUNT');
+    require(_currentExposure + assetAmount <= _exposureCap, 'EXOGENOUS_ASSET_EXPOSURE_TOO_HIGH');
+
+    _currentExposure += uint128(assetAmount);
+    _accruedFees += fee.toUint128();
+    IERC20(UNDERLYING_ASSET).safeTransferFrom(originator, address(this), assetAmount);
+
+    IGhoToken(GHO_TOKEN).mint(address(this), grossAmount);
+    IGhoToken(GHO_TOKEN).transfer(receiver, ghoBought);
+    // TRIGGER ERROR: invalid transfer of GHO amount to GSM Converter (msg.sender)
+    IGhoToken(GHO_TOKEN).transfer(msg.sender, 1);
```

Comment:
> you can just have this one broken function that you call in your test

This one I think was harder to simulate bc it needed to happen within the flow of the converter, rather than within the test. 

> effectively testing a contract out of scope of our repo, no

I don't think it is out of scope of the repo, but perhaps out of scope of the converter because it is the GSM implementation. Do you think from the POV of the converter I can assume that the GSM will function properly and leave the require statement coverage uncovered that checks the correct amount of tokens are returned from GSM?

Reply:
I think you can leave this mock in here, but can't you reduce the code? I think you probably left in some functions that are unused in the tests involving this mock

Reply:
yes I think so, let me work on that